cmd/alertmanager/main.go

```diff
@@ -317,9 +323,28 @@ func run() int {
 		go peer.Settle(ctx, *gossipInterval*10)
 	}
 
-	alerts, err := mem.NewAlerts(context.Background(), marker, *alertGCInterval, logger)
-	if err != nil {
-		level.Error(logger).Log("err", err)
+	var alerts provider.Alerts
+	if *alertStorageProvider == "mem" {
+		alerts, err = mem.NewAlerts(context.Background(), marker, *alertGCInterval, logger)
+		if err != nil {
+			level.Error(logger).Log("err", err)
+			return 1
+		}
+	} else if *alertStorageProvider == "etcd" {
+		etcdAlerts, err := etcd.NewAlerts(context.Background(), marker, *alertGCInterval, logger, *alertEtcdEndpoints, *alertEtcdPrefix)
+		if err != nil {
+			level.Error(logger).Log("err", err)
+			return 1
+		}
+		alerts = etcdAlerts
+		go func() {
```

Comment:
Q: So this runs concurrently?  What happens to AM during the very first sleep? Maybe a channel is necessary here?

Reply:
During that sleep, AM is receiving events.  I think those run loops will fail out when the context gets canceled.  I'm pretty new to go.  Seems to work fine in the tests